wgengine/magicsock/magicsock.go

```diff
@@ -3310,6 +3312,10 @@ func (c *RebindingUDPConn) WriteBatch(msgs []ipv6.Message, flags int) (int, erro
 
 		n, err = bw.WriteBatch(msgs[start:], flags)
 		if err != nil {
+			if errors.Is(err, errNoSys) && !oldKernel {
```

Comment:
it would be nice if we could detect this in `func upgradePacketConn()` somehow.

Reply:
We could try to parse the kernel version like I just did in https://github.com/golang/sys/commit/2204b6615fb84fc69a9e5d37b2d508ddf544dcea but I kinda hate that... vendor kernels are weird cherry-pick messes of different things. I prefer to just feature test at runtime and see what we get.

But that means we'd need to make some dummy system calls on upgrade.

Maybe just parse very broadly: "Kernel version start with '2'? Sorry, no fast mode for you!"

Reply:
e.g.
```patch
diff --git a/wgengine/magicsock/magicsock.go b/wgengine/magicsock/magicsock.go
index 5c42867d..415726d7 100644
--- a/wgengine/magicsock/magicsock.go
+++ b/wgengine/magicsock/magicsock.go
@@ -37,6 +37,7 @@
        "tailscale.com/derp/derphttp"
        "tailscale.com/disco"
        "tailscale.com/health"
+       "tailscale.com/hostinfo"
        "tailscale.com/ipn/ipnstate"
        "tailscale.com/logtail/backoff"
        "tailscale.com/net/connstats"
@@ -3208,6 +3209,14 @@ type RebindingUDPConn struct {
 func upgradePacketConn(p nettype.PacketConn, network string) nettype.PacketConn {
        uc, ok := p.(*net.UDPConn)
        if ok && runtime.GOOS == "linux" && (network == "udp4" || network == "udp6") {
+               // recvmmsg/sendmmsg were added in 2.6.33 but we support down to 2.6.32
+               // for old NAS devices. See https://github.com/tailscale/tailscale/issues/6807.
+               // As a cheap heuristic: if the Linux kernel starts with "2", just consider
+               // it too old for the fast paths. Nobody who cares about performance runs such
+               // ancient kernels.
+               if strings.HasPrefix(hostinfo.GetOSVersion(), "2") {
+                       return p
+               }
                // Non-Linux does not support batch operations. x/net will fall back to
                // recv/sendmsg, but not all platforms have recv/sendmsg support. Keep
                // this simple for now.
```

Reply:
I wonder if a zero len `msgs` will get down into the syscall. Though, that may be fragile/break later if it does. Alternatively, we could move `oldKernel` up into `RebindingUDPConn` so we don't have to evaluate on every call to Read/WriteBatch.

Reply:
Just tested this `"2"` variant.  Both versions work, but the `"2"` variant is less intrusive. PTAL.